pkg/metrics/sketch_series.go

```diff
@@ -78,6 +79,118 @@ func (sl SketchSeriesList) MarshalJSON() ([]byte, error) {
 	return reqBody.Bytes(), err
 }
 
+// MarshalSplitCompress uses the stream compressor to marshal and compress sketch series payloads.
+// If a compressed payload is larger than the max, a new payload will be generated. This method returns a slice of
+// compressed protobuf marshaled gogen.SketchPayload objects. gogen.SketchPayload is not directly marshaled - instead
+// it's contents are marshaled individually, packed with the appropriate protobuf metadata, and compressed in stream.
+// The resulting payloads (when decompressed) are binary equal to the result of marshaling the whole object at once.
+func (sl SketchSeriesList) MarshalSplitCompress() ([]*[]byte, error) {
+	input := bytes.NewBuffer(make([]byte, 0, 1024))
+	output := bytes.NewBuffer(make([]byte, 0, 1024))
+
+	// The Metadata field of gogen.SketchPayload is never written to - so pack an empty metadata as the footer
+	footer := []byte{0x12, 0}
+
+	compressor, e := stream.NewCompressor(input, output, []byte{}, footer, []byte{})
+	if e != nil {
+		return nil, e
+	}
+	payloads := []*[]byte{}
+	precompressionBuf := make([]byte, 1024)
+
+	dsl := make([]gogen.SketchPayload_Sketch_Dogsketch, 1)
+	for _, ss := range sl {
+		if len(ss.Points) > cap(dsl) {
+			dsl = make([]gogen.SketchPayload_Sketch_Dogsketch, len(ss.Points))
+		}
+
+		for i, p := range ss.Points {
+			b := p.Sketch.Basic
+			k, n := p.Sketch.Cols()
+			dsl[i] = gogen.SketchPayload_Sketch_Dogsketch{
+				Ts:  p.Ts,
+				Cnt: b.Cnt,
+				Min: b.Min,
+				Max: b.Max,
+				Avg: b.Avg,
+				Sum: b.Sum,
+				K:   k,
+				N:   n,
+			}
+		}
+
+		sketch := gogen.SketchPayload_Sketch{
+			Metric:      ss.Name,
+			Host:        ss.Host,
+			Tags:        ss.Tags,
+			Dogsketches: dsl[:len(ss.Points)],
+		}
+
+		// Pack the protobuf metadata - see SketchPayload.MarshalTo in agent_payload.pb.go for reference.
+		i := 0
+		// Magic number that occurs before the varint encoding
+		precompressionBuf[i] = 0xa
+		i++
+		i = encodeVarintAgentPayload(precompressionBuf, i, uint64(sketch.Size()))
+
+		// Resize the pre-compression buffer if needed
+		totalItemSize := sketch.Size() + i
+		if totalItemSize > cap(precompressionBuf) {
+			precompressionBuf = append(precompressionBuf, make([]byte, totalItemSize-cap(precompressionBuf))...)
+			precompressionBuf = precompressionBuf[:cap(precompressionBuf)]
+		}
+
+		// Marshal the sketch to the precompression buffer after the metadata
+		_, e := sketch.MarshalTo(precompressionBuf[i:])
+		if e != nil {
+			return nil, e
+		}
+
+		// Compress the protobuf metadata and the marshaled sketch
+		e = compressor.AddItem(precompressionBuf[:totalItemSize])
+		if e == stream.ErrItemTooBig || e == stream.ErrPayloadFull {
```

Comment:
in which cases can `ErrItemTooBig` happen?

We should track it with telemetry

Reply:
If the item itself is larger than the max allowed item size defined in the compressor it won't be accepted. [The comment here sums it up pretty well](https://github.com/DataDog/datadog-agent/blob/e943fbbd951465afb4cfe0a7a703d0f2a1e527c9/pkg/serializer/jsonstream/compressor.go#L106). 

FWIW - `split.Payloads` (the old logic) doesn't really handle this case nicely - it appears that split will hit is maximum split loop bound and just drop the payload.

+1 I'll add telemetry here - and ill also explicitly handle this case when we re-pack before the next iteration because now that I look closer this can be handled in a better way.